File: frontend/src/util/canvas/calculateGradeFromOutcomes.js
const naGrade = { grade: 'N/A' };

export default (outcomeRollups, userId) => {
  const courseIds = Object.keys(outcomeRollups);

  const grades = {};

  courseIds.forEach(i => {
    const rollup = outcomeRollups[i];

    // array of outcomes for a specified class
    const scores = rollup[0].scores.map(s => s.score);

    if (scores.length < 1) {
      grades[i] = { grade: 'N/A' };
      return;
    }

    grades[i] = getGradeFromOutcomes(scores);
  });

  return grades;
};

function getGradeFromOutcomes(outcomes) {
  // outcomes, ex: [1,2,3,4]

  // what is 75% of outcomes.length
  const outcomesOverMinNeeded = Math.floor((75 * outcomes.length) / 100);

  // desc
  const sortedOutcomes = outcomes.sort((a, b) => b - a);
  // 75% of outcomes
  const countedOutcomes = sortedOutcomes.slice(0, outcomesOverMinNeeded);

  // lowest counted outcome
  const lowestCountedOutcome = countedOutcomes[countedOutcomes.length - 1];
  // lowest outcome (overall)
  const lowestOutcome = sortedOutcomes[sortedOutcomes.length - 1];

  // map lowest counted outcome to grade map
  for (let i = 0; i < gradeMap.length; i++) {
    // current available grade
    const grade = gradeMap[i][0];
    // not really max, but what all counted outcomes must be over
    const max = gradeMap[i][1][0];
    // what all outcomes must be over
    const min = gradeMap[i][1][1];

    // lowest outcome over minimum?
    if (lowestOutcome < min) {
      // one outcome is below min for this grade
      continue;
    }

    // counted outcomes must be above max, too
    if (lowestCountedOutcome < max) {
      // 75% of outcomes are not over max
      continue;
    }

    // student is eligible for this grade
    return {
      grade,
      sortedOutcomes,
      countedOutcomes,
      lowestCountedOutcome,
      lowestOutcome
    };
  }

  // we should never reach this point-- it would require a negative outcome
}

export const gradeMap = [
  ['A', [3.3, 3]],
  ['A-', [3.3, 2.5]],
  ['B+', [2.6, 2.5]],
  ['B', [2.6, 0]],
  ['B-', [2.6, 0]],
  ['C', [2.2, 0]],
  ['I', [0, 0]]
];

export const gradeMapByGrade = {
  A: { rank: 6, max: 3.3, min: 3 },
  'A-': { rank: 5, max: 3.3, min: 2.5 },
  'B+': { rank: 4, max: 2.6, min: 2.5 },
  B: { rank: 3, max: 2.6, min: 0 },
  'B-': { rank: 2, max: 2.6, min: 0 },
  C: { rank: 1, max: 2.2, min: 0 },
  I: { rank: 0, max: 0, min: 0 }
};

File: backend/src/db/canvas/grades/calculate_grade_from_outcomes.go
package grades

import (
	"math"
	"sort"
)

type Grade struct {
	// the Grade, like A or B
	Grade string
	// how "good" the Grade is, higher is better
	Rank uint8
	// 75% = most; aka max
	MostAbove float64
	// aka min
	AllAbove float64
}

// gradeMap is a slice of possible grades
var gradeMap = []Grade{
	{"A", 6, 3.3, 3},
	{"A-", 5, 3.3, 2.5},
	{"B+", 4, 2.6, 2.2},
	{"B", 3, 2.6, 1.8},
	{"B-", 2, 2.6, 1.5},
	{"C", 1, 2.2, 1.5},
	{"I", 0, 0, 0},
}

// CalculateGradeFromOutcomeScores calculates a Grade from outcome scores (os)
func CalculateGradeFromOutcomeScores(os []float64) Grade {
	// what is 75% of len(s)
	outcomesOverMinNeeded := int(math.Floor(float64(75*len(os)) / float64(100)))

	// float64 outcome scores
	sortedOutcomes := sort.Float64Slice(os)
	// sorts in place
	sort.Sort(sort.Reverse(sortedOutcomes))
	// sortedOutcomes are now sorted

	// first 75% of outcomes
	countedOutcomes := sortedOutcomes[:outcomesOverMinNeeded]

	// if there is only one graded outcome in a class, that outcome is counted
	// this also fixes an array[-1] bug
	if len(countedOutcomes) < 1 {
		countedOutcomes = sortedOutcomes
	}

	lowestCountedOutcome := countedOutcomes[len(countedOutcomes)-1]

	// overall
	lowestOutcome := sortedOutcomes[len(sortedOutcomes)-1]

	// default to an I
	finalGrade := gradeMap[6]

	for _, v := range gradeMap {
		// lowest outcome is over minimum (AllAbove)
		if v.AllAbove > lowestOutcome {
			continue
		}

		// lowestCountedOutcome must be above MostAbove
		if v.MostAbove > lowestCountedOutcome {
			continue
		}

		// student qualifies for this Grade
		if v.Rank > finalGrade.Rank {
			finalGrade = v
		}
	}

	return finalGrade
}
